### PR TITLE
Build elements only when they are ready - @mkhatib

### LINKED ISSUE #3081: Reimplement `isReadyToBuild` callback inside of AMP Runtime
Currently, AMP Runtime relies on `isReadyToBuild` callback to decide when the element can be built. Ideally, an `<amp-element>` should be built as soon as `</amp-element>` is parsed. However, there's no callback of this kind available to us. Instead, we want to rely on the "next element parsed" signal. It will work as following:

1. On resource add (`Resources.add()` method) we will NOT call `Resource.build()`, but instead we will push the resource to the `openElements[]` buffer.
2. On resource add (`Resources.add()`) we will scan all elements in the `openElements[]` buffer to see if they have the next element in the document order. "Next element in the document order" is defined as: (a) the element itself has the `nextSibling`, (b) the parent has the `nextSibling`, (c) the parent of parent (and so on) has the `nextSibling`. If we see the "next element in the document order", we remove this element from the `openElements` buffer and call `build()` callback on it.
3. On `DOMReady`, we call `build()` on all remaining elements in the `openElements[]` buffer and no longer buffer new elements. All post-`DOMReady` added resources will be built immediately (as is the case today).
4. Thus we will immediately stop using `BaseElement.isReadyToBuild()` callback and eventually drop it from the `BaseElement` class and its sub-classes.

### PR DESCRIPTION
**Pull request by mkhatib**
So, this only checks if the element has its `nextSibling` and does not traverse the parent tree. Waiting on reply on #3081 to understand why do we need to traverse the parents.

If this looks good, I can start adding tests as well!

Fixes #3081

### CODE REVIEW

File: src/service/resources-impl.js [hunk 1]
```diff
@@ -387,15 +391,42 @@ export class Resources {
...
+   * Builds resources that are ready to be built.
+   * @private
+   */
+  buildReadyResources_() {
+    const closedElements = [];
+    this.openElements_.forEach(el => {
+      if (this.documentReady_ || el.nextSibling) {
+        const resource = this.getResourceForElement(el);
+        resource.build(this.forceBuild_);
+        closedElements.push(el);
```
Comment (mkhatib) [inline src/service/resources-impl.js]: Probably could also just store the indexes instead of the elements themselves to avoid checking the index in the cleanup loop.

File: src/custom-element.js [hunk 2]
```diff
@@ -457,7 +457,7 @@ export function createAmpElementProto(win, name, implementationClass) {
       return true;
     }
     dev.assert(this.isUpgraded(), 'Cannot build unupgraded element');
-    if (!force && !this.implementation_.isReadyToBuild()) {
```
Comment (dvoytenko) [inline src/custom-element.js]: `force` argument has to go completely.

File: src/service/resources-impl.js [hunk 3]
```diff
@@ -141,6 +141,9 @@ export class Resources {
     /** @private {!Array<!Function>} */
     this.deferredMutates_ = [];
 
+    /** @private {!Array<!Element>} */
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: Let's do `?` so that we can reset it after we no longer collecting open elements. This way it'd be easy to catch any regressions if anyone is trying to use this array after.

File: src/service/resources-impl.js [hunk 4]
```diff
@@ -387,15 +391,42 @@ export class Resources {
     this.resources_.push(resource);
 
     if (this.isRuntimeOn_) {
-      // Try to immediately build element, it may already be ready.
-      resource.build(this.forceBuild_);
-      this.schedulePass();
+      if (this.documentReady_) {
+        // Try to immediately build element, it may already be ready.
+        resource.build(this.forceBuild_);
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: No more `forceBuild` argument.

File: src/service/resources-impl.js [hunk 5]
```diff
@@ -387,15 +391,42 @@ export class Resources {
...
+        this.schedulePass();
+      } else {
+        // Track open elements
+        this.openElements_.push(element);
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: It'd be easier to track if you add `resource` instead of `element` - but it's up to you.

File: src/service/resources-impl.js [hunk 6]
```diff
@@ -387,15 +391,42 @@ export class Resources {
...
+      }
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: Also, don't forget to clean this array on resource removal - just as we did with other queues.

File: src/service/resources-impl.js [hunk 7]
```diff
@@ -387,15 +391,42 @@ export class Resources {
...
   }
 
   /**
+   * Builds resources that are ready to be built.
+   * @private
+   */
+  buildReadyResources_() {
+    const closedElements = [];
+    this.openElements_.forEach(el => {
+      if (this.documentReady_ || el.nextSibling) {
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: So, as mentioned in the bug, `el.nextSubling` is not enough - you also have to go all the way up the parent chain to search their `nextSibling` as well.

File: src/service/resources-impl.js (hunk 1 above)
Comment (dvoytenko) [inline src/service/resources-impl.js]: Yes, that'd be better, but with reverse traversal for removal.

File: src/service/resources-impl.js (hunk 5 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: That sounds good. Mind if I call this `pendingBuildResources_` instead? It really signifies what this really is used for better than `openElements` or `openResources`?

File: src/service/resources-impl.js (hunk 4 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done. Should I go ahead and remove the usages for `isReadyToBuild` across AMP? Or should I do that in a separate PR?

File: src/custom-element.js (hunk 2 above)
Comment (mkhatib) [inline src/custom-element.js]: Done.

File: src/service/resources-impl.js (hunk 3 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done.

File: src/service/resources-impl.js (hunk 6 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done.

File: src/service/resources-impl.js (hunk 7 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done.

File: src/service/resources-impl.js (hunk 1 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done.

**Comment by mkhatib** [comment]
PTAL.

File: src/dom.js [hunk 8]
```diff
@@ -383,3 +383,19 @@ export function getDataParamsFromAttributes(element, opt_computeParamNameFunc) {
...
 }
+
+
+/**
+ * Whether the element has been already fully parsed by the browser.
+ * @param {!Element} element
+ * @return {boolean}
+ */
+export function isElementFullyParsed(element) {
+  let currentElement = element;
```
Comment (cramforce) [inline src/dom.js]: Document the algorithm.

File: src/dom.js [hunk 9]
```diff
@@ -383,3 +383,19 @@ export function getDataParamsFromAttributes(element, opt_computeParamNameFunc) {
...
   return params;
 }
+
+
+/**
+ * Whether the element has been already fully parsed by the browser.
+ * @param {!Element} element
+ * @return {boolean}
+ */
+export function isElementFullyParsed(element) {
```
Comment (cramforce) [inline src/dom.js]: Add a test

File: src/service/resources-impl.js [hunk 10]
```diff
@@ -387,15 +388,40 @@ export class Resources {
...
   }
 
   /**
+   * Builds resources that are ready to be built.
+   * @private
+   */
+  buildReadyResources_() {
+    const builtResourcesIndexes = [];
+    this.pendingBuildResources_.forEach((resource, index) => {
+      if (this.documentReady_ || isElementFullyParsed(resource.element)) {
```
Comment (cramforce) [inline src/service/resources-impl.js]: Any chance this logic can be tested?

Comment (mkhatib) [inline src/service/resources-impl.js]: Yeah I'll add some tests, just wanted to make sure this is the right path to take.

File: src/service/resources-impl.js [hunk 11]
```diff
@@ -387,15 +388,40 @@ export class Resources {
...
     dev.fine(TAG_, 'element added:', resource.debugid);
   }
 
   /**
+   * Builds resources that are ready to be built.
+   * @private
+   */
+  buildReadyResources_() {
+    const builtResourcesIndexes = [];
+    this.pendingBuildResources_.forEach((resource, index) => {
```
Comment (jridgewell) [inline src/service/resources-impl.js]: If this is as busy as I think it is, we should opt for a `for` loop to gain back some speed. Inlining the `#splice` (and subtracting 1 from the `index`) should be faster, too.

File: src/service/resources-impl.js [hunk 12]
```diff
@@ -387,15 +388,40 @@ export class Resources {
...
+      if (this.documentReady_ || isElementFullyParsed(resource.element)) {
+        resource.build();
+        builtResourcesIndexes.push(index);
+      }
+    });
+
+    // Remove resources that we just built from pending resources.
+    for (let i = builtResourcesIndexes.length; i > 0; i--) {
```
Comment (jridgewell) [inline src/service/resources-impl.js]: `builtResourcesIndexes.length - 1`, or else `i` will be out of bounds on the first run.

File: src/service/resources-impl.js [hunk 13]
```diff
@@ -387,15 +388,40 @@ export class Resources {
...
     }
 
     dev.fine(TAG_, 'element added:', resource.debugid);
   }
 
   /**
+   * Builds resources that are ready to be built.
+   * @private
+   */
+  buildReadyResources_() {
```
Comment (jridgewell) [inline src/service/resources-impl.js]: Won't this do a full loop of every `pendingBuildResources_` every time we add an element? Should we debounce this to prevent a ton of looping?

File: src/service/resources-impl.js [hunk 14]
```diff
@@ -387,15 +388,40 @@ export class Resources {
     this.resources_.push(resource);
 
     if (this.isRuntimeOn_) {
-      // Try to immediately build element, it may already be ready.
-      resource.build(this.forceBuild_);
-      this.schedulePass();
+      if (this.documentReady_) {
```
Comment (jridgewell) [inline src/service/resources-impl.js]: Can we check if `element` is already parsed? That'll save pushing and splicing from `pendingBuildResources_` (and a lot of looping).

File: src/dom.js [hunk 15]
```diff
@@ -383,3 +383,19 @@ export function getDataParamsFromAttributes(element, opt_computeParamNameFunc) {
...
+  let currentElement = element;
+  do {
+    if (currentElement.nextSibling) {
```
Comment (erwinmombay) [inline src/dom.js]: `currentElement.nextElementSibling`

Comment (jridgewell) [inline src/dom.js]: That may actually slow things down. A `TextNode` also implies that the element has been parsed, not just an `Element`.

Comment (erwinmombay) [inline src/dom.js]: good point, wasn't aware we were just doing a check.

Comment (jridgewell) [inline src/dom.js]: Using `#nextSibling` also has the benefit of whitespace nodes:

```html
<div>
  <amp-element></amp-element>
</div>
```
There are 2 `TextNodes` in that, and importantly one is is directly after the `amp-element` closing tag. With `#nextSibling`, we may never traverse the parents at all.

File: src/service/resources-impl.js (hunk 13 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Debouncing might have a negative effect here. Since elements are probably added pretty fast, debouncing might just delay our execution till the whole document was parsed - this would defeats the purpose. I am not sure if there's an easy way to test this out, but maybe a throttling approach would work better where it won't delay the execution but instead limits how frequent this is being called.

@dvoytenko any idea if this is a concern or is it safe to just call it on every element add?

Comment (jridgewell) [inline src/service/resources-impl.js]: Or, we can break when we find the first unparsed element (assuming elements are added in the same order they parse).

File: src/service/resources-impl.js (hunk 11 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: I really hate manipulating both the array and the index iterator while looping. I think collecting the indexes to delete and then deleting them in reverse is slightly safer and easier to read.

But if we expect it'll improve performance happy to do it. @dvoytenko you ok with doing the splice and index manipulation instead of two separate loops?

File: src/service/resources-impl.js (hunk 12 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: good catch 😊

File: src/service/resources-impl.js (hunk 13 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Yes that assumption is correct, but not sure if breaking when we find the first unparsed element is accurate - for example:

```javascript
<amp-element>
  <amp-child1>
  <amp-child2>
  <amp-child3>
</amp-element>
```

Breaking at the first unparsed element will always break on the `amp-element` because it has not yet fully parsed although some of its children probably already did - so they're all delayed until its parent is fully parsed.

File: src/custom-element.js [hunk 16]
```diff
@@ -435,31 +435,19 @@ export function createAmpElementProto(win, name, implementationClass) {
...
    *
-   * @param {boolean} force Whether or not force the build.
    * @return {boolean}
    * @final @this {!Element}
    */
-  ElementProto.build = function(force) {
+  ElementProto.build = function() {
     this.assertNotTemplate_();
     if (this.isBuilt()) {
       return true;
```
Comment (dvoytenko) [inline src/custom-element.js]: Returnring `true` is also not logical now. We just need to return or fail here.

File: src/custom-element.js [hunk 17]
```diff
@@ -435,31 +435,19 @@ export function createAmpElementProto(win, name, implementationClass) {
...
-   *
-   * If the "force" argument is "true", the element performs build regardless
-   * of what {@link BaseElement.isReadyToBuild} would return.
-   *
    * Returned value indicates whether or not build has been performed.
    *
    * This method can only be called on a upgraded element.
    *
-   * @param {boolean} force Whether or not force the build.
    * @return {boolean}
```
Comment (dvoytenko) [inline src/custom-element.js]: ditto on return value.

File: src/service/resources-impl.js (hunk 13 above)
Comment (jridgewell) [inline src/service/resources-impl.js]: Or or, we can use binary search (What?! This may be the first time I've ever used a search algorithm in real life)! Find the first non-parse element, everything before is parse, everything after isn't.

File: src/service/resources-impl.js (hunk 14 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: This seems like a good idea, the only thing is we would't be calling `buildReadyResources_` until a non-parsed element appear or documentReady kicks off, I think this might delay some of the elements that were already added to `pendingBuildResources`. No?

Comment (dvoytenko) [inline src/service/resources-impl.js]: It'd be good. The only question I'd ask: the current approach is nice because it builds elements in the order they are registered. I don't think the order matters, but let's double-check if we may have some negatives.

File: src/service/resources-impl.js (hunk 13 above)
Comment (dvoytenko) [inline src/service/resources-impl.js]: Not necessarily true. They can also be added as engine discovers custom element binding.

File: src/service/resources-impl.js [hunk 18]
```diff
@@ -1493,6 +1518,8 @@ export class Resources {
       });
       this.requestsChangeSize_ = this.requestsChangeSize_.filter(
           request => request.resource != resource);
+      this.pendingBuildResources_ = this.pendingBuildResources_.filter(
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: Currently you are storing elements in the `pendingBuildResources_`, but here you check for resources. You have to decide which one you store.

File: src/service/resources-impl.js [hunk 19]
```diff
@@ -1628,16 +1655,15 @@ export class Resource {
   /**
    * Requests the resource's element to be built. See {@link AmpElement.build}
    * for details.
-   * @param {boolean} force
    * @return {boolean}
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: Ditto on response value.

File: src/dom.js [hunk 20]
```diff
@@ -383,3 +383,19 @@ export function getDataParamsFromAttributes(element, opt_computeParamNameFunc) {
   }
   return params;
 }
+
+
+/**
+ * Whether the element has been already fully parsed by the browser.
```
Comment (dvoytenko) [inline src/dom.js]: I don't think this is a good name since we can't really judge "parsed" in all different cases. But we can name it something like "hasNextNodeInDocumentOrder" or such that directly explains what it does.

File: src/service/resources-impl.js (hunk 18 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Oh I thought i've updated that reference, I'll be storing resources.

File: src/service/resources-impl.js (hunk 19 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: You saying we want to drop the return value? It seems to be still useful in this context, but maybe not in the `ElementProto.build` one.

Comment (mkhatib) [inline src/service/resources-impl.js]: Nevermind, I got what you were saying. Just return instead of return false/true.

File: src/service/resources-impl.js (hunk 11 above)
Comment (dvoytenko) [inline src/service/resources-impl.js]: I'm ok to do this in one for-loop, as long as we don't find any negatives from reverse order build. I can't think of any problems offhand.

File: src/service/resources-impl.js (hunk 13 above)
Comment (dvoytenko) [inline src/service/resources-impl.js]: So, I'd say we definitely can't assume any order of added elements for many reasons. We have to scan the whole array all the time. On the plus side we know that most of times this array will have only 1-2 elements.

File: src/service/resources-impl.js (hunk 19 above)
Comment (dvoytenko) [inline src/service/resources-impl.js]: Well, in this case you are correct. We can preserve return value here since it's actually meaningful in this case. However, it's definitely no longer meaningful for `ElementProto.build` and should be removed there.

File: src/service/resources-impl.js (hunk 11 above)
Comment (jridgewell) [inline src/service/resources-impl.js]: Two loops may be ok, but we should definitely change this to a for loop.

As for the merged possibility, we can do forward iteration, just `index--` when we splice something.

File: src/service/resources-impl.js (hunk 19 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done

File: src/service/resources-impl.js (hunk 18 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done

File: src/service/resources-impl.js (hunk 12 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done

File: src/dom.js (hunk 15 above)
Comment (mkhatib) [inline src/dom.js]: Ack.

File: src/dom.js (hunk 8 above)
Comment (mkhatib) [inline src/dom.js]: Done

File: src/dom.js (hunk 9 above)
Comment (mkhatib) [inline src/dom.js]: Done

File: src/dom.js (hunk 20 above)
Comment (mkhatib) [inline src/dom.js]: Done

File: src/custom-element.js (hunk 16 above)
Comment (mkhatib) [inline src/custom-element.js]: Done

File: src/custom-element.js (hunk 17 above)
Comment (mkhatib) [inline src/custom-element.js]: Done

**Comment by mkhatib** [comment]
Still need to fix the tests and add some tests for the new build flow.

File: src/service/resources-impl.js (hunk 19 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: hah. I've kinda removed it from here, should I bring it back? I tried to see where do we actually use it, couldn't find a place where we do check the returned value, but in the future, it might be useful.

Comment (dvoytenko) [inline src/service/resources-impl.js]: No. It's not really used. So it's fine.

File: src/custom-element.js (hunk 17 above)
Comment (dvoytenko) [inline src/custom-element.js]: Still en-route? Still seeing it.

File: src/service/resources-impl.js (hunk 4 above)
Comment (dvoytenko) [inline src/service/resources-impl.js]: `isReadyToBuild` - please remove in the follow-up PR. Please create a bug for that.

File: src/service/resources-impl.js (hunk 11 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Just curious why for-loop is better?

Comment (jridgewell) [inline src/service/resources-impl.js]: Avoids a function allocation, avoids edge cases with holes that `#forEach` has to deal with. In general, SPEED. 🐎 🏇

File: src/service/resources-impl.js (hunk 4 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Filed #3196

File: src/service/resources-impl.js (hunk 11 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Ok, so one loop this is how it looks like:
```javascript
  buildReadyResources_() {
    for (let i = 0; i < this.pendingBuildResources_.length;  i++) {
      const resource = this.pendingBuildResources_[i];
      if (this.documentReady_ || hasNextNodeInDocumentOrder(resource.element)) {
        resource.build();
        // Resource is built remove it from the pending list and step back
        // one in the index to account for the removed item.
        this.pendingBuildResources_.splice(i--, 1);
      }
    }
  }
```

File: src/service/resources-impl.js (hunk 13 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Ack.

File: src/service/resources-impl.js (hunk 14 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Dima, also is my concern in the comment above valid? Can a pending resource be delayed unnecessary because we haven't encounter another non-fully parsed element yet? 

If not I can't think of a negative out-of-order building. So we can do this if there isn't a concern of delaying some resources.

File: src/custom-element.js (hunk 17 above)
Comment (mkhatib) [inline src/custom-element.js]: Yeah it's gone now. Don't know why it was still showing.

File: src/service/resources-impl.js (hunk 19 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Ack.

File: src/service/resources-impl.js (hunk 18 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: @dvoytenko now that I actually think of it, we really shouldn't clean this up when the element unload! This is for building not for layouting so we should always keep it in order for the element to be built otherwise if we clean this and the user come back to the page the element will never be built if it was pending!

Let me know if I am missing something.

Comment (mkhatib) [inline src/service/resources-impl.js]: I've dropped the cleanup here. Let me know if I got this wrong.

File: src/service/resources-impl.js (hunk 11 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done.

File: src/service/resources-impl.js (hunk 10 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Added tests.

Done.

**Comment by mkhatib** [comment]
Added tests fixed the comments. @dvoytenko PTAL

File: src/service/resources-impl.js (hunk 14 above)
Comment (dvoytenko) [inline src/service/resources-impl.js]: I agree. I don't see real negatives from order here. So, let's ignore that question.

File: src/service/resources-impl.js (hunk 18 above)
Comment (dvoytenko) [inline src/service/resources-impl.js]: This is not to do with load/unload, right? Basically, if the resource has been removed from DOM - we no longer need to wait to build it.

File: src/service/resources-impl.js [hunk 21]
```diff
@@ -141,6 +138,9 @@ export class Resources {
     /** @private {!Array<!Function>} */
     this.deferredMutates_ = [];
 
+    /** @private {?Array<!Element>} */
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: Please update jsdoc. Now it's `?Array<!Resource>`, right?

File: src/service/resources-impl.js (hunk 14 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Does this mean we do want to add the check and skip adding to pending queue?

File: src/service/resources-impl.js (hunk 18 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Thanks for clarifying this on chat, correct, this needs to happen when the element is removed not unloaded. Also during chat we discussed the extra fix to cleanup layouting tasks for a resource when it's removed as well as unloaded.

Comment (mkhatib) [inline src/service/resources-impl.js]: Both Done.

File: src/service/resources-impl.js (hunk 21 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done

File: src/service/resources-impl.js [hunk 22]
```diff
@@ -1494,6 +1518,13 @@ export class Resources {
       this.requestsChangeSize_ = this.requestsChangeSize_.filter(
           request => request.resource != resource);
     }
+
+    if (resource.getState() == ResourceState_.NOT_BUILT && opt_removePending) {
+      const pendingIndex = this.pendingBuildResources_.indexOf(resource);
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: `pendingBuildResources_` can be `null` here - so need to check that it's not.

File: src/service/resources-impl.js [hunk 23]
```diff
@@ -407,6 +430,7 @@ export class Resources {
     if (index != -1) {
       this.resources_.splice(index, 1);
     }
+    this.cleanupTasks_(resource, /** removePending */ true);
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: nit: it should be `/* removePending */` since it's not a jsdoc.

**Comment by dvoytenko** [comment]
LGTM, but please address the latest two comments.

File: src/service/resources-impl.js (hunk 22 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done

File: src/service/resources-impl.js (hunk 23 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done.

**Comment by dvoytenko** [comment]
yay!

File: src/service/resources-impl.js [hunk 24]
```diff
@@ -876,7 +899,7 @@ export class Resources {
     for (let i = 0; i < this.resources_.length; i++) {
       const r = this.resources_[i];
       if (r.getState() == ResourceState_.NOT_BUILT) {
-        r.build(this.forceBuild_);
+        r.build();
```
Comment (lannka) [inline src/service/resources-impl.js]: @mkhatib Do we need to check `hasNextNode` here before calling the build?
Or is this whole method `discoverWork_` should never be invoked before document ready?

We hit a bug here: #4397
